File: Softeer/우물 안 개구리.py
from sys import stdin
from collections import deque

n, m = map(int, stdin.readline().split())
weight = [0] + list(map(int, stdin.readline().split()))
graph = [[] for _ in range(n + 1)]
cant = set()

for i in range(m):
    a, b = map(int, stdin.readline().split())
    graph[a].append(b)
    graph[b].append(a)

def bfs(s):
    flag = True
    for v in graph[s]:
        if weight[v] > weight[s]:
            cant.add(s)
            flag = False
            break
        elif weight[v] < weight[s]:
            cant.add(v)
            continue
        elif weight[v] == weight[s]:
            flag = False
            cant.add(v)
            cant.add(s)
            break
    return flag
                
res = 0

for node in range(1, n + 1):
    if len(graph[node]) == 0:
        res += 1
        continue
    else:
        if node not in cant:
            if bfs(node):
                res += 1
print(res)

File: Softeer/우물 안 개구리.js
var input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
var [n, m] = input[0].split(' ').map(Number);
var weights = [0, ...input[1].split(' ').map(Number)];
var graph = new Array(n + 1).fill().map(() => []);
input.splice(0, 2);

var cant = new Set();

for(let i = 0; i < m; i++){
    var [a, b] = input[i].split(' ').map(Number); 
    graph[a].push(b);
    graph[b].push(a);
}

const bfs = (v) => {
    for(node of graph[v]){
        if(weights[node] < weights[v]){
            cant.add(node);
            continue;
        } else if (weights[node] > weights[v]){
            cant.add(v);
            return false;
        } else {
            cant.add(v);
            cant.add(node);
            return false;
        }
    }
    return true;
}

var res = 0;
for(let node = 1; node <= n; node++){
    if(graph[node].length === 0){
        res += 1;
        continue;
    } else {
        if(cant.has(node) === false){
            if(bfs(node)){res += 1;}
        }
    }
}
console.log(res)

